File: DeepSpace-Java/src/deepspace/GameUniverse.java
package deepspace;
import java.util.ArrayList;

public class GameUniverse {
    private static int WIN = 10;

    private int turns;
    private Dice dice;
    private int currentStationIndex;
    private EnemyStarShip currentEnemy;
    private SpaceStation currentStation;
    private GameStateController gameState;
    private ArrayList<SpaceStation> spaceStations = new ArrayList<>();

    private Boolean haveSpaceCity = false;

    GameUniverse() {
        gameState = new GameStateController();
        dice      = new Dice();
        turns     = 0;
    }

//
// ───────────────────────────────────────────────────────── METODOS PRIVADOS ─────
//

    private void createSpaceCity() {

    }

    private void makeStationEfficient () {

    }

//
// ──────────────────────────────────────────────────────────────────── MOUNTS ─────
//

    public void mountShieldBooster(int i){
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.mountShieldBooster(i);
    }

    public void mountWeapon(int i){
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.mountWeapon(i);
    }

//
// ───────────────────────────────────────────────────────────────── DISCARDS ─────
//

    public void discardHangar () {
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.discardHangar();
    }

    public void discardShieldBooster(int i) {
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.discardShieldBooster(i);
    }

    public void discardShieldBoosterInHangar(int i) {
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.discardShieldBoosterInHangar(i);
    }


    public void discardWeapon (int i) {
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.discardWeapon(i);
    }

    public void discardWeaponInHangar(int i) {
        if (gameState.getState() == GameState.INIT || gameState.getState() == GameState.AFTERCOMBAT)
            currentStation.discardWeaponInHangar(i);
    }

//
// ─────────────────────────────────────────────────────────────── INTERFACES ─────
//

    public GameState getState() {
        return gameState.getState();
    }

    public GameUniverseToUI getUIversion() {
        return new GameUniverseToUI(currentStation, currentEnemy);
    }

//
// ──────────────────────────────────────────────────────────── OTROS METODOS ─────
//

    public void init (ArrayList<String> names) {
        GameState state = gameState.getState();
        if (state == GameState.CANNOTPLAY){
            spaceStations = new ArrayList<>();
            CardDealer dealer = CardDealer.getInstance();

            for (int i = 0; i < names.size(); ++i) {
                SuppliesPackage supplies = dealer.nextSuppliesPackage();
                SpaceStation station = new SpaceStation(names.get(i), supplies);
                spaceStations.add(station);

                int nh = dice.initWithNHangars();
                int nw = dice.initWithNWeapons();
                int ns = dice.initWithNShields();

                Loot lo = new Loot(0, nw, ns, nh, 0);

                station.setLoot(lo);
            }

            currentStationIndex = dice.whoStarts(names.size());
            currentStation = spaceStations.get(currentStationIndex);
            currentEnemy = dealer.nextEnemy();

            gameState.next(turns, spaceStations.size());
        }
    }

    public boolean nextTurn () {
        GameState state = gameState.getState();
        if (state == GameState.AFTERCOMBAT){
            Boolean stationState = currentStation.validState();
            if(stationState){
                currentStationIndex=(currentStationIndex+1)%spaceStations.size();
                currentStation = spaceStations.get(currentStationIndex);
                currentStation.cleanUpMountedItems();
                CardDealer dealer = CardDealer.getInstance();
                currentEnemy = dealer.nextEnemy();
                gameState.next(turns,spaceStations.size());
                return true;
            }
            return false;
        }
        return false;
    }

    public boolean haveAWinner(){
        return currentStation.getNMedals() == WIN;
    }

    public CombatResult combat () {
        GameState state = gameState.getState();
        if ((state == GameState.BEFORECOMBAT) || (state == GameState.INIT)){
            return combat(currentStation, currentEnemy);
        }
        return CombatResult.NOCOMBAT;
    }

    CombatResult combat (SpaceStation station, EnemyStarShip enemy) {
        Boolean enemyWins;
        CombatResult combatResult;
        GameCharacter ch = dice.firstShot();
        if (ch == GameCharacter.ENEMYSTARSHIP){
            float fire = enemy.fire();
            ShotResult result = station.receiveShot(fire);
            if (result == ShotResult.RESIST){
                fire = station.fire();
                result = enemy.receiveShot(fire);
                enemyWins = (result == ShotResult.RESIST);
            }
            else{
                enemyWins = true;
            }
        }
        else{
            float fire = station.fire();
            ShotResult result = enemy.receiveShot(fire);
            enemyWins = (result == ShotResult.RESIST);
        }

        if (enemyWins){
            float s = station.getSpeed();
            Boolean moves = dice.spaceStationMoves(s);
            if (!moves){
                Damage damage = enemy.getDamage();
                station.setPendingDamage(damage);
                combatResult = CombatResult.ENEMYWINS;
            }
            else{
                station.move();
                combatResult = CombatResult.STATIONSCAPES;
            }
        }
        else{
            Loot aLoot = enemy.getLoot();
            station.setLoot(aLoot);
            combatResult = CombatResult.STATIONWINS;
        }

        gameState.next(turns, spaceStations.size());
        return combatResult;
    }

    public String toString () {
        return      "\t-> turnos: " + turns
                + "\n\t-> Dado: " + dice.toString()
                + "\n\t-> Estación actual: " + currentStation.toString()
                + "\n\t-> Enemigo actual: " + currentEnemy.toString()
                + "\n\t-> Estado:" + gameState.toString();
    }
}

File: DeepSpace-Ruby/lib/GameUniverse.rb
# encoding:utf-8
require_relative "GameUniverseToUI"
require_relative "GameStateController"
require_relative "Dice"
require_relative "CombatResult"
require_relative "GameCharacter"
require_relative "ShotResult"
require_relative "SpaceStation"
require_relative "CardDealer"
require_relative "EnemyStarShip"
require_relative "PowerEfficientSpaceStation"
require_relative "BetaPowerEfficientSpaceStation"
require_relative "SpaceCity"
require_relative "Transformation"

module Deepspace
class GameUniverse
    @@WIN = 10

    def initialize
        @turns               = 0
        @currentStationIndex = -1

        @dice          = Dice.new
        @spaceStations = Array.new
        @gameState     = GameStateController.new

        @currentEnemy
        @currentStation
        @haveSpaceCity = false
    end


#
# ────────────────────────────────────────────────────────────────── COMBATS ─────
#

    def combatGo (station, enemy)
        ch = @dice.firstShot
        combatResult = CombatResult::ENEMYWINS

        # Determinar quién gana: si enemigo o estación. Depende de quién empiece
        if ch == GameCharacter::ENEMYSTARSHIP
            fire   = enemy.fire
            result = station.receiveShot(fire)

            if result == ShotResult::RESIST
               fire      = station.fire
               result    = enemy.receiveShot(fire)
               enemyWins = (result == ShotResult::RESIST)
            else
                enemyWins = true
            end
        else
            fire      = station.fire
            result    = enemy.receiveShot(fire)
            enemyWins = (result == ShotResult::RESIST)
        end


        if enemyWins
            s     = station.getSpeed
            moves = @dice.spaceStationMoves(s)

            if !moves
                damage = enemy.damage
                station.setPendingDamage( damage )
                combatResult = CombatResult::ENEMYWINS
            else
                station.move
                combatResult = CombatResult::STATIONESCAPES
            end
        else
            aLoot = enemy.loot
            setransforma = station.setLoot( aLoot )
            if setransforma == Transformation::GETEFFICIENT
                makeStationEfficient
                combatResult = CombatResult::STATIONWINSANDCONVERTS
            elsif setransforma == Transformacion::SPACECITY
                createSpaceCity
                combatResult = CombatResult::STATIONWINSANDCONVERTS
            else
                combatResult = CombatResult::STATIONWINS
            end
        end

        @gameState.next(@turns, @spaceStations.length)

        combatResult
    end


    def combat
        state = @gameState.state

        if state == GameState::BEFORECOMBAT || state == GameState::INIT
            combatGo(@currentStation, @currentEnemy)
        else
            CombatResult::NOCOMBAT
        end
    end

#
# ───────────────────────────────────────────────────────────────── DISCARDS ─────
#

    def discardHangar
        if @gameState.state == GameState::INIT or @gameState.state == GameState::AFTERCOMBAT
            @currentStation.discardHangar
        end
    end

    def discardShieldBooster(i)
        if @gameState.state == GameState::INIT or @gameState.state == GameState::AFTERCOMBAT
            @currentStation.discardShieldBooster(i)
        end
    end

    def discardShieldBoosterInHangar(i)
        if @gameState.state == GameState::INIT or @gameState.state == GameState::AFTERCOMBAT
            @currentStation.discardShieldBoosterInHangar(i)
        end
    end

    def discardWeapon(i)
        if @gameState.state == GameState::INIT or @gameState.state == GameState::AFTERCOMBAT
            @currentStation.discardWeapon(i)
        end
    end

    def discardWeaponInHangar(i)
        if @gameState.state == GameState::INIT or @gameState.state == GameState::AFTERCOMBAT
            @currentStation.discardWeaponInHangar(i)
        end
    end

#
# ─────────────────────────────────────────────────────────────────── MOUNTS ─────
#

    def mountShieldBooster(i)
        if @gameState.state == GameState::INIT or @gameState.state == GameState::AFTERCOMBAT
            @currentStation.mountShieldBooster(i)
        end
    end

    def mountWeapon(i)
        if @gameState.state == GameState::INIT or @gameState.state == GameState::AFTERCOMBAT
            @currentStation.mountWeapon(i)
        end
    end

#
# ──────────────────────────────────────────────────────────────────── OTROS ─────
#

    def state
        @gameState.state
    end

    def getUIversion
        GameUniverseToUI.new(@currentStation, @currentEnemy)
    end

    def haveAWinner
        return @currentStation.nMedals >= @@WIN
    end

    def init (names) #String[]
        state = @gameState.state

        if state == GameState::CANNOTPLAY
            dealer = CardDealer.instance # Debería ser atributo de instancia?

            for i in 0..names.size-1
                # Crear y preparar naves para los jugadores
                supplies = dealer.nextSuppliesPackage

                station = SpaceStation.new(names[i], supplies)

                nh = @dice.initWithNHangars
                nw = @dice.initWithNWeapons
                ns = @dice.initWithNShields

                lo = Loot.new(0, nw, ns, nh, 0)
                station.setLoot(lo)

                @spaceStations.push(station)
            end

            @currentStationIndex = @dice.whoStarts(names.size)
            @currentStation      = @spaceStations[@currentStationIndex]
            @currentEnemy        = dealer.nextEnemy

            @gameState.next(@turns, @spaceStations.size)
        end
    end

    def nextTurn
        state = @gameState.state
        return_value = false

        if state == GameState::AFTERCOMBAT
            if @currentStation.validState
                @currentStationIndex = (@currentStationIndex + 1) % @spaceStations.size
                @turns += 1

                @currentStation = @spaceStations[@currentStationIndex]

                @currentStation.cleanUpMountedItems

                dealer = CardDealer.instance

                @currentEnemy = dealer.nextEnemy

                @gameState.next(@turns, @spaceStations.size)
            end
        end


    end

private

    def createSpaceCity
        if @haveSpaceCity == false
            vector = Arrayn.new
            @spaceStations.each{ |station| 
                if station != @currentStation
                    vector << station
                 end
            }

            @currentStation = SpaceCity.new(@currentStation, vector)
            @spaceStations[@currentStationIndex] = @currentStation
            @haveSpaceCity = true
        end
    end

    def makeStationEfficient
        @currentStation = PowerEfficientSpaceStation.new(@currentStation)
        if @dice.extraEfficiency
            @currentStation = BetaPowerEfficientSpaceStation.new(@currentStation)
        end

        @spaceStations[@currentStationIndex] = @currentStation
    end            

public    

    def to_s
        getUIversion.to_s
    end
end
end
